src/SpeedJumpIrm.sol

```diff
@@ -120,43 +124,39 @@ contract AdaptativeCurveIrm is IIrm {
             return (uint256(_curve(INITIAL_RATE_AT_TARGET, err)), INITIAL_RATE_AT_TARGET);
         } else {
             // Note that the speed is assumed constant between two interactions, but in theory it increases because of
             // interests. So the rate will be slightly underestimated.
             int256 speed = ADJUSTMENT_SPEED.wMulDown(err);
-
             // market.lastUpdate != 0 because it is not the first interaction with this market.
             // Safe "unchecked" cast because block.timestamp - market.lastUpdate <= block.timestamp <= type(int256).max.
             int256 elapsed = int256(block.timestamp - market.lastUpdate);
             int256 linearAdaptation = speed * elapsed;
-            int256 adaptationMultiplier = MathLib.wExp(linearAdaptation);
-            // endRateAtTarget is bounded between MIN_RATE_AT_TARGET and MAX_RATE_AT_TARGET.
-            int256 endRateAtTarget =
-                startRateAtTarget.wMulDown(adaptationMultiplier).bound(MIN_RATE_AT_TARGET, MAX_RATE_AT_TARGET);
-            int256 endBorrowRate = _curve(endRateAtTarget, err);
-
-            // Then we compute the average rate over the period.
-            // Note that startBorrowRate is defined in the computations below.
-            // avgBorrowRate = 1 / elapsed * ∫ startBorrowRate * exp(speed * t) dt between 0 and elapsed
-            //               = startBorrowRate * (exp(linearAdaptation) - 1) / linearAdaptation
-            //               = (endBorrowRate - startBorrowRate) / linearAdaptation
-            // And for linearAdaptation around zero: avgBorrowRate ~ startBorrowRate = endBorrowRate.
-            // Also, when it is the first interaction (rateAtTarget = 0).
-            int256 avgBorrowRate;
-            if (linearAdaptation == 0) {
-                avgBorrowRate = endBorrowRate;
-            } else {
-                int256 startBorrowRate = _curve(startRateAtTarget, err);
-                avgBorrowRate = (endBorrowRate - startBorrowRate).wDivDown(linearAdaptation);
+
+            // We want to approximate an average of the rate over the period [0, T]:
+            // avg = 1/T ∫_0^T curve(startRateAtTarget * exp(speed * x), err) dx
+            // We approximate the integral with a Riemann sum (steps of length T/N):
+            // avg ~= 1/T Σ_i=1^N curve(startRateAtTarget * exp(speed * T/N * i), err) * T / N
+            //     ~= Σ_i=1^N curve(startRateAtTarget * exp(linearVariation/N * i), err) / N
+            // curve is linear in startRateAtTarget, so:
+            //     ~= curve(Σ_i=1^N startRateAtTarget * exp(linearVariation/N * i), err) / N
+            //     ~= curve(Σ_i=1^N startRateAtTarget * exp(linearVariation/N * i) / N, err)
+            int256 sumRateAtTarget;
+            int256 step = linearAdaptation / N_STEPS;
+            // Stops after N_STEPS - 1, the last one is done with endRateAtTarget.
+            for (int256 k = 1; k < N_STEPS; k++) {
```

Comment:
Let's do the left Riemann for err > 0 (even if it means not reusing the endRateAtTarget, sorry)